### PR TITLE
[transformer] Allow for skipping stream synch

### PR DESCRIPTION
**Pull request by crcrpar**
Optionally disable stream synchronization after batched p2p communication

exported from nvcr.io/nvidia/pytorch:22.09-py3 container with some test cases

cc @eqy @Aidyn-A @ptrblck @Fuzzkatt

### CODE REVIEW

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 1]
```diff
@@ -310,6 +323,58 @@ def test_learning_pipelining_without_interleaving_ucc_for_p2p(self):
     def test_inference_pipelining_without_interleaving_ucc_for_p2p(self):
         self._test_hybrid_backends(True)
 
+    @unittest.skipUnless(SUPPORT_ASYNC_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_inference_pipelining_without_interleaving_async_batch_isend_irecv(self):
```
Comment (Aidyn-A) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py]: Suggested name: `test_learning_pipelining_without_interleaving_async_batch_isend_irecv`

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 2]
```diff
@@ -310,6 +323,58 @@ def test_learning_pipelining_without_interleaving_ucc_for_p2p(self):
...
+            pipeline_model_parallel_world_size=None,
+            virtual_pipeline_model_parallel_size=None,
+            async_comm=False,
+            default_backend=None,
+            p2p_backend=None,
+            sync_batch_comm=False,
+        )
+
+    @unittest.skipUnless(SUPPORT_ASYNC_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_inference_pipelining_without_interleaving_async_batch_isend_irecv_infrerence(self):
```
Comment (Aidyn-A) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:333]: Perhaps `test_inference_pipelining_without_interleaving_async_batch_isend_irecv` would do it.

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 3]
```diff
@@ -310,6 +323,58 @@ def test_learning_pipelining_without_interleaving_ucc_for_p2p(self):
...
+            pipeline_model_parallel_world_size=None,
+            virtual_pipeline_model_parallel_size=None,
+            async_comm=False,
+            default_backend=None,
+            p2p_backend=None,
+            sync_batch_comm=False,
+        )
+
+    @unittest.skipUnless(SUPPORT_ASYNC_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_inference_pipelining_with_interleaving_async_batch_isend_irecv(self):
```
Comment (Aidyn-A) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:337]: Here: `test_learning_pipelining_with_interleaving_async_batch_isend_irecv`

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 4]
```diff
@@ -310,6 +323,58 @@ def test_learning_pipelining_without_interleaving_ucc_for_p2p(self):
...
+            pipeline_model_parallel_world_size=None,
+            virtual_pipeline_model_parallel_size=None,
+            async_comm=False,
+            default_backend=None,
+            p2p_backend=None,
+            sync_batch_comm=False,
+        )
+
+    @unittest.skipUnless(SUPPORT_ASYNC_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_inference_pipelining_with_interleaving_async_batch_isend_irecv_inference(self):
```
Comment (Aidyn-A) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:341]: And lastly: `test_inference_pipelining_with_interleaving_async_batch_isend_irecv`

**Review by Aidyn-A — APPROVED** [review]
Sorry, one of my latest PRs created a confusion is test naming. I prefer using `test_learning` and `test_inference` for more (subjectively) convenient usage when running the tests. So I made several suggestions for them. Everything else looks good to me :+1:

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 5]
```diff
@@ -310,6 +329,38 @@ def test_learning_pipelining_without_interleaving_ucc_for_p2p(self):
...
+    def test_learning_pipelining_without_interleaving_skyp_sync_after_batch_isend_irecv(self):
+        self.test_learning_pipelining_without_interleaving(sync_batch_comm=False)
+
+    @unittest.skipUnless(CAN_SKIP_SYNC_AFTER_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_inference_pipelining_without_interleaving_skip_sync_after_batch_isend_irecv(self):
+        self.test_inference_pipelining_without_interleaving(sync_batch_comm=False)
+
+    @unittest.skipUnless(CAN_SKIP_SYNC_AFTER_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_learning_async_pipelining_without_interleaving_skip_sync_after_batch_isend_irecv(self):
+        self.test_learning_async_pipelining_without_interleaving(sync_batch_comm=True)
```
Comment (Aidyn-A) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:342]: Shouldn't be here `sync_batch_comm=False`?

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 6]
```diff
@@ -310,6 +329,38 @@ def test_learning_pipelining_without_interleaving_ucc_for_p2p(self):
...
+
+    @unittest.skipUnless(CAN_SKIP_SYNC_AFTER_BATCH_ISEND_IRECV, "Requires https://github.com/pytorch/pytorch/pull/82450")
+    def test_inference_async_pipelining_without_interleaving_skip_sync_after_batch_isend_irecv(self):
+        self.test_inference_async_pipelining_without_interleaving(sync_batch_comm=True)
```
Comment (Aidyn-A) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:346]: And here `sync_batch_comm=False` as well?

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py (hunk 5 above)
Comment (crcrpar) [reply tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py]: yeah, thanks. this pr is a well-balanced task which made me consider as a routine... :/

**Review by Aidyn-A — APPROVED** [review]

**Comment by crcrpar** [comment]
on DGX A100, `python tests/L0/run_test.py --include run_transformer` encountered out of memory error and interleaving cases failed, while `(python|pytest) tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py` not. So this PR looks unrelated.

**Pull request merged**